components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/request/PostAuthnMissingChallengeQuestionsHandler.java

```diff
@@ -0,0 +1,411 @@
+            return ChallengeQuestionManager.getInstance().getAllChallengeQuestions
+                    (tenantDomain);
+        } catch (IdentityRecoveryServerException e) {
+            log.error("Identity recovery server error occurred for user:" + user.getUserName(), e);
+            return null;
+        }
+    }
+
+    /**
+     * Set the challenge questions for the user by calling ChallengeQuestionManger.
+     *
+     * @param user                 Authenticated User.
+     * @param userChallengeAnswers Array of UserChallengeAnswer
+     */
+    private void setChallengeQuestionAnswers(User user, UserChallengeAnswer[] userChallengeAnswers) {
+
+        try {
+            ChallengeQuestionManager.getInstance().setChallengesOfUser(user, userChallengeAnswers);
+        } catch (IdentityRecoveryException e) {
+            log.error("Unable to save challenge question answers", e);
+        }
+    }
+
+    /**
+     * Returns array of UserChallengeAnswer from the servlet request parameters from the front-end
+     *
+     * @param servletRequest HTTP Servlet Request.
+     * @return Array of UserChallengeAnswer.
+     */
+    private UserChallengeAnswer[] retrieveChallengeQuestionAnswers(HttpServletRequest servletRequest,
+                                                                   List<ChallengeQuestion> list) {
+        Map<String, String> questionsMap = new HashMap<>();
+        Map<String, String> answersMap = new HashMap<>();
+        List<UserChallengeAnswer> questionsAndAnswers = new ArrayList<>();
+        List<String> paramNamesList = new ArrayList<>();
+
+        Enumeration<String> paramNames = servletRequest.getParameterNames();
+        while (paramNames.hasMoreElements()) {
+            paramNamesList.add(paramNames.nextElement());
+        }
+
+        for (String requestParam : paramNamesList) {
+            if (requestParam.contains(SELECTED_CHALLENGE_QUESTION_PREFIX)) {
+                String question = servletRequest.getParameter(requestParam);
+                String questionSetID = requestParam.replace(SELECTED_CHALLENGE_QUESTION_PREFIX, "");
+                questionsMap.put(questionSetID, question);
+
+            } else if (requestParam.contains(CHALLENGE_QUESTION_ANSWER_PREFIX)) {
+                String answer = servletRequest.getParameter(requestParam);
+                String answerSetID = requestParam.replace(CHALLENGE_QUESTION_ANSWER_PREFIX, "");
+                answersMap.put(answerSetID, answer);
+            }
+        }
+        for (String questionKey : questionsMap.keySet()) {
+            String challengeQuestion = questionsMap.get(questionKey);
+            for (ChallengeQuestion question : list) {
+                if (question.getQuestion().equals(challengeQuestion)) {
+                    UserChallengeAnswer questionAndAnswer = new UserChallengeAnswer();
+                    questionAndAnswer.setQuestion(question);
+                    questionAndAnswer.setAnswer(answersMap.get(questionKey));
```

Comment:
dont we need to check whether answersMap.get(questionKey) exists before setting as the answer?

Reply:
Also can there be unanswered challenge questions at this phase?